frontend/javascripts/oxalis/model/sagas/save_saga.js

```diff
@@ -52,55 +62,144 @@ import compactUpdateActions from "oxalis/model/helpers/compaction/compact_update
+          yield* fork(
+            compressBucketAndAddToUndoBatch,
+            zoomedBucketAddress,
+            bucketData,
+            currentVolumeAnnotationBatch,
+          ),
+        );
+      } else if (finishAnnotationStrokeAction) {
+        yield* join([...pendingCompressions]);
+        bucketAlreadyInUndoState.clear();
+        undoStack.push({ type: "volume", data: currentVolumeAnnotationBatch });
+        currentVolumeAnnotationBatch = [];
+        pendingCompressions = [];
+      }
+      // Clear the redo stack when a new action is executed.
+      redoStack.splice(0);
+      if (undoStack.length > UNDO_HISTORY_SIZE) {
+        undoStack.shift();
       }
     } else if (undo) {
-      if (undoStack.length) {
+      if (undoStack.length > 0 && undoStack[undoStack.length - 1].type === "skeleton") {
         previousAction = null;
-        redoStack.push(prevTracing);
-        const newTracing = undoStack.pop();
-        yield* put(setTracingAction(newTracing));
-        yield* put(centerActiveNodeAction());
-      } else {
-        Toast.info(messages["undo.no_undo"]);
       }
+      yield* call(
+        applyStateOfStack,
+        undoStack,
+        redoStack,
+        prevTracingMaybe,
+        messages["undo.no_undo"],
+      );
     } else if (redo) {
-      if (redoStack.length) {
-        undoStack.push(prevTracing);
-        const newTracing = redoStack.pop();
-        yield* put(setTracingAction(newTracing));
-        yield* put(centerActiveNodeAction());
-      } else {
-        Toast.info(messages["undo.no_redo"]);
+      if (redoStack.length > 0 && redoStack[redoStack.length - 1].type === "skeleton") {
+        previousAction = null;
       }
+      yield* call(
+        applyStateOfStack,
+        redoStack,
+        undoStack,
+        prevTracingMaybe,
+        messages["undo.no_redo"],
+      );
     }
     // We need the updated tracing here
-    prevTracing = yield* select(state => enforceSkeletonTracing(state.tracing));
+    prevTracingMaybe = yield* select(state => state.tracing.skeleton);
+  }
+}
+
+function* getSkeletonTracingToUndoState(
+  skeletonUserAction: SkeletonTracingAction,
+  prevTracing: SkeletonTracing,
+  previousAction: ?SkeletonTracingAction,
+): Saga<?SkeletonUndoState> {
+  const curTracing = yield* select(state => enforceSkeletonTracing(state.tracing));
+  if (curTracing !== prevTracing) {
+    if (shouldAddToUndoStack(skeletonUserAction, previousAction)) {
+      return { type: "skeleton", data: prevTracing };
+    }
+  }
+  return null;
+}
+
+function* compressBucketAndAddToUndoBatch(
+  zoomedBucketAddress: Vector4,
+  bucketData: BucketDataArray,
+  undoBatch: VolumeAnnotationBatch,
+): Saga<void> {
+  const bucketDataAsByteArray = new Uint8Array(
+    bucketData.buffer,
+    bucketData.byteOffset,
+    bucketData.byteLength,
+  );
+  const compressedBucketData = yield* call(byteArrayToLz4Array, bucketDataAsByteArray, true);
+  if (compressedBucketData != null) {
```

Comment:
Removing the if check seems to be not possible. Flow thinks that `compressedBucketData` might be undefined. 

My guess is, that this might be because we use a webworker for this and if the webworker fails, undefined will be returned :man_shrugging: